File: le-liaison/src/main/python/liaison/QueryColumnVDBImpl.py

#
# $LastChangedBy$
# $LastChangedDate$
# $Rev$
#

import re
from .exceptions import MaudeStringError
from .QueryColumn import QueryColumn
from .ExpressionVDBImpl import *


class QueryColumnVDBImpl(QueryColumn):

## This is the default constructor

    def __init__(self, name, expression, approved_usage = None, display_name = None,
                 category = None, statistical_type = None, tags = None,
                 fundamental_type = None, description = None,
                 display_discretization = None):

        self.initFromValues(name, expression, approved_usage, display_name, category,
                            statistical_type, tags, fundamental_type,
                            description, display_discretization)


## This is the pythonic way to create multiple constructors.  The call is
## c = QueryColumnVDBImpl.InitFromDefn( defn )

    @classmethod
    def initFromDefn(cls, defn):

        name                   = None
        expression             = None

        md = {}
        md['ApprovedUsage']    = None
        md['DisplayName']      = None
        md['Category']         = None
        md['StatisticalType']  = None
        md['Tags']             = None
        md['FundamentalType']  = None
        md['Description']      = None
        md['DisplayDiscretizationStrategy'] = None

        c_no_metadata = re.search( '^SpecQueryNamedFunctionExpression\(ContainerElementName\(\"(\w*)\"\), (LatticeFunction.*)\)$', defn )

        if c_no_metadata:
            name = c_no_metadata.group(1)
            expression = ExpressionVDBImplFactory.Create( c_no_metadata.group(2) )

        c_fcnbndry = re.search( '^SpecQueryNamedFunctionEntityFunctionBoundary$', defn )

        if c_fcnbndry:
            name = 'EntityFunctionBoundary'
            expression = ExpressionVDBImplFactory.Create( 'SpecQueryNamedFunctionEntityFunctionBoundary' )
        
        c_has_metadata = re.search( '^SpecQueryNamedFunctionMetadata\(SpecQueryNamedFunctionExpression\(ContainerElementName\(\"(\w*)\"\), (LatticeFunction.*)\), SpecExtractDetails\(\((.*)\)\)\)$', defn )

        if c_has_metadata:
            name = c_has_metadata.group(1)
            expression = ExpressionVDBImplFactory.Create( c_has_metadata.group(2) )
            sed = c_has_metadata.group(3)

            while True:
                metadata_is_extracted = False
                
                for md_name in md:
                    md_found = re.search( '^SpecExtractDetail\(\"{0}\", \"(.*?)\"\)(, SpecExtractDetail\(.*|$)'.format(md_name), sed )
                    if md_found:
                        md[md_name] = md_found.group(1).replace('\\"','"')
                        sed = md_found.group(2)[2:]
                        metadata_is_extracted = True
                        break

                    md_found = re.search( '^SpecExtractDetail\(\"{0}\", StringList\(.*?\"(.*?)\"\)\)(, SpecExtractDetail\(.*|$)'.format(md_name), sed )
                    if md_found:
                        md[md_name] = md_found.group(1).replace('\\"','"')
                        sed = md_found.group(2)[2:]
                        metadata_is_extracted = True
                        break

                if not metadata_is_extracted:
                    raise MaudeStringError( 'Unsupported metadata type: {0}'.format(sed) )

                if sed == '':
                    break

        if expression is None:
            raise MaudeStringError( defn )
        
        return cls(name, expression, md['ApprovedUsage'], md['DisplayName'], md['Category'],
                   md['StatisticalType'], md['Tags'], md['FundamentalType'], md['Description'],
                   md['DisplayDiscretizationStrategy'])


    def definition(self):

        if self.getName() == 'EntityFunctionBoundary':
            return 'SpecQueryNamedFunctionEntityFunctionBoundary'

        mdsep = ''
        md = ''
        if self.getDisplayName() is not None:
            md += mdsep + 'SpecExtractDetail("DisplayName", "{0}")'.format( self.getDisplayName().replace('"','\\"') )
            mdsep = ', '
        if self.getDescription() is not None:
            md += mdsep + 'SpecExtractDetail("Description", "{0}")'.format( self.getDescription().replace('"','\\"') )
            mdsep = ', '
        if self.getApprovedUsage() is not None:
            md += mdsep + 'SpecExtractDetail("ApprovedUsage", "{0}")'.format( self.getApprovedUsage().replace('"','\\"') )
            mdsep = ', '
        if self.getFundamentalType() is not None:
            md += mdsep + 'SpecExtractDetail("FundamentalType", "{0}")'.format( self.getFundamentalType().replace('"','\\"') )
            mdsep = ', '
        if self.getStatisticalType() is not None:
            md += mdsep + 'SpecExtractDetail("StatisticalType", "{0}")'.format( self.getStatisticalType().replace('"','\\"') )
            mdsep = ', '
        if self.getTags() is not None:
            md += mdsep + 'SpecExtractDetail("Tags", "{0}")'.format( self.getTags().replace('"','\\"') )
            mdsep = ', '
        if self.getDisplayDiscretization() is not None:
            md += mdsep + 'SpecExtractDetail("DisplayDiscretizationStrategy", "{0}")'.format( self.getDisplayDiscretization().replace('"','\\"') )
            mdsep = ', '
        if self.getCategory() is not None:
            md += mdsep + 'SpecExtractDetail("Category", "{0}")'.format( self.getCategory().replace('"','\\"') )
            mdsep = ', '

        sqnf =  'SpecQueryNamedFunctionExpression('
        sqnf +=   'ContainerElementName("{0}")'.format( self.getName() )
        sqnf += ', '+ self.getExpression().definition()
        sqnf += ')'

        if md == '':
            return sqnf

        sqnfm =  'SpecQueryNamedFunctionMetadata('
        sqnfm += sqnf
        sqnfm += ', SpecExtractDetails(('
        sqnfm += md
        sqnfm +=   '))'
        sqnfm += ')'
        
        return sqnfm

File: le-liaison/src/main/java/com/latticeengines/liaison/service/impl/QueryColumnVDBImpl.java
package com.latticeengines.liaison.service.impl;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.latticeengines.liaison.exposed.exception.DefinitionException;
import com.latticeengines.liaison.exposed.service.QueryColumn;

public class QueryColumnVDBImpl extends QueryColumn {

    private static final Set<String> mdTypes = new HashSet<String>(Arrays.asList("ApprovedUsage",
            "DisplayName", "Category", "StatisticalType", "Tags", "FundamentalType", "Description",
            "DisplayDiscretizationStrategy"));

    private static final Pattern pattern_sqnf_no_md = Pattern
            .compile("^SpecQueryNamedFunctionExpression\\(ContainerElementName\\(\"(\\w*)\"\\), (LatticeFunction.*)\\)$");
    private static final Pattern pattern_sqnf_no_md_alt = Pattern
            .compile("^SpecQueryNamedFunctionMetadata\\(SpecQueryNamedFunctionExpression\\(ContainerElementName\\(\"(\\w*)\"\\), (LatticeFunction.*)\\), SpecExtractDetails\\(empty\\)\\)$");
    private static final Pattern pattern_sqnf_fcnbndry = Pattern
            .compile("^SpecQueryNamedFunctionEntityFunctionBoundary$");
    private static final Pattern pattern_sqnf_has_md = Pattern
            .compile("^SpecQueryNamedFunctionMetadata\\(SpecQueryNamedFunctionExpression\\(ContainerElementName\\(\"(\\w*)\"\\), (LatticeFunction.*)\\), SpecExtractDetails\\(\\((.*)\\)\\)\\)$");

    private static final String tmpl_sed_single = "^SpecExtractDetail\\(\"%s\", \"(.*?)\"\\)(, SpecExtractDetail\\(.*|$)";
    private static final String tmpl_sed_list = "^SpecExtractDetail\\(\"%s\", StringList\\(.*?\"(.*?)\"\\)\\)(, SpecExtractDetail\\(.*|$)";

    public QueryColumnVDBImpl(String definition) throws DefinitionException {
        super(definition);
    }

    public QueryColumnVDBImpl(QueryColumnVDBImpl other) {
        super(other);
    }

    public void initFromDefinition(String defn) throws DefinitionException {

        String name = "NOT SET";
        String expression = "NOT SET";
        Map<String, String> metadata = new HashMap<>();
        boolean isMatched = Boolean.FALSE;

        if (!isMatched) {
            // Case (1): There is no metadata attached
            Matcher c_no_md = pattern_sqnf_no_md.matcher(defn);
            if (c_no_md.matches()) {
                name = c_no_md.group(1);
                expression = c_no_md.group(2);
                isMatched = Boolean.TRUE;
            }
        }

        if (!isMatched) {
            // Case (2): There is an empty metadata object
            Matcher c_no_md_alt = pattern_sqnf_no_md_alt.matcher(defn);
            if (c_no_md_alt.matches()) {
                name = c_no_md_alt.group(1);
                expression = c_no_md_alt.group(2);
                isMatched = Boolean.TRUE;
            }
        }

        if (!isMatched) {
            // Case (3): The entity-attribute boundary
            Matcher c_fcnbndry = pattern_sqnf_fcnbndry.matcher(defn);
            if (c_fcnbndry.matches()) {
                name = "EntityFunctionBoundary";
                expression = "SpecQueryNamedFunctionEntityFunctionBoundary";
                isMatched = Boolean.TRUE;
            }
        }

        if (!isMatched) {
            // Case (4): There is metadata attached
            Matcher c_has_md = pattern_sqnf_has_md.matcher(defn);
            if (c_has_md.matches()) {
                name = c_has_md.group(1);
                expression = c_has_md.group(2);
                isMatched = Boolean.TRUE;

                String sed = c_has_md.group(3);

                while (Boolean.TRUE) {
                    boolean isExtracted = Boolean.FALSE;
                    for (String mdType : mdTypes) {
                        Matcher md_found = Pattern.compile(String.format(tmpl_sed_single, mdType))
                                .matcher(sed);
                        if (md_found.matches()) {
                            metadata.put(mdType, md_found.group(1).replace("\\\"", "\""));
                            sed = md_found.group(2);
                            isExtracted = Boolean.TRUE;
                            break;
                        }
                        md_found = Pattern.compile(String.format(tmpl_sed_list, mdType)).matcher(
                                sed);
                        if (md_found.matches()) {
                            metadata.put(mdType, md_found.group(1).replace("\\\"", "\""));
                            sed = md_found.group(2);
                            isExtracted = Boolean.TRUE;
                            break;
                        }
                    }

                    if (!isExtracted) {
                        throw new DefinitionException(String.format(
                                "Maude definition (SpecExtractDetail) cannot be interpretted: %s",
                                sed));
                    }

                    if (sed.equals("")) {
                        break;
                    }
                    sed = sed.substring(2);
                }
            }
        }

        if (expression.equals("NOT SET")) {
            throw new DefinitionException(String.format(
                    "Maude definition (SpecQueryNamedFunction) cannot be interpretted: %s", defn));
        }

        initFromValues(name, expression, metadata);
    }

    public String definition() {

        if (getName().equals("EntityFunctionBoundary")) {
            return "SpecQueryNamedFunctionEntityFunctionBoundary";
        }

        String mdsep = "";
        StringBuilder mddefn = new StringBuilder(1000);
        Map<String, String> metadata = getMetadata();
        for (Map.Entry<String, String> entry : metadata.entrySet()) {
            mddefn.append(mdsep).append(
                    String.format("SpecExtractDetail(\"%s\", \"%s\")", entry.getKey(), entry
                            .getValue().replace("\"", "\\\"")));
            mdsep = ", ";
        }

        StringBuilder sqnf = new StringBuilder(1000);
        sqnf.append("SpecQueryNamedFunctionExpression(");
        sqnf.append(String.format("ContainerElementName(\"%s\")", getName()));
        sqnf.append(", " + getExpression());
        sqnf.append(")");

        if (mddefn.toString().equals("")) {
            return sqnf.toString();
        }

        StringBuilder sqnfm = new StringBuilder(1000);
        sqnfm.append("SpecQueryNamedFunctionMetadata(");
        sqnfm.append(sqnf.toString());
        sqnfm.append(", SpecExtractDetails((");
        sqnfm.append(mddefn.toString());
        sqnfm.append("))");
        sqnfm.append(")");

        return sqnfm.toString();
    }
}
